[3.x] Fix Polygon2D skinned bounds (for culling)

lawnjelly opened the pull request:
The bound Rect2 was previously incorrect because bone transforms need to be applied to verts in bone space (or rigged space), rather than local space. This was previously resulting in skinned Polygon2Ds being incorrectly culled.

Also adds `debug_canvas_item_get_rect()` function to `VisualServer` for debugging / verification / possible unit testing.

Fixes #63047 (2nd issue reported by @hothex)

### Testing project
Contains skinned Polygon2Ds, and displays bounding boxes in green.

[Skinned_Polygon2D_CullingTest.zip](https://github.com/godotengine/godot/files/11209674/Skinned_Polygon2D_CullingTest.zip)

https://user-images.githubusercontent.com/21999379/230062627-0a2f5a60-ecaf-413d-a696-35fbb44d6afb.mp4

## Explanation
There were a number of difficulties to overcome in generating the bounding rect in local space, compared to the skinning in the renderer. In particular in the renderer it was easier to find the relative global transform between the `Polygon2D` and the `Skeleton`. In the `get_rect()` call, this information was not available.

* The _relative transform_ is important to deal with the case where there is an offset between the `Polygon2D` and the `Skeleton`. When they are exactly aligned, there is no relative transform to deal with and things are simpler - but users tend to rig with relative transforms.

The `final_transform` in the canvas item is not so useful because it is only calculated _after_ rendering, and there is a chicken and egg problem where we need to know the global transform of the `Polygon2D` _before_ rendering, because in the current paradigm the global transforms are not stored in the visual server, only calculated on the fly during rendering. Additionally the `final_transform` contained also the camera transform, and thus could not be compared with the `Skeleton` global transform.

I explored two options to finding the relative transform:
1) Send the global transform of `Polygon2D`s from scene side to `VisualServer`. This would entail keeping this updated whenever the `Polygon2D` moved, which is quite a few updates. Additionally the relative transform would need to be calculated each time.
2) By checking the numbers I noticed that the relative transform typically doesn't change once the bones are rigged to a skeleton. This means we had the opportunity to calculate the relative transform as a one off, and send it to the `VisualServer`. This is the approach I have taken in this PR.

## Notes
* The most expensive part (vertex operations) is now precalculated as a one off when the rigging is established. This means that at runtime, in most cases it transforms and merges one or two bones, which should be pretty cheap.
* This error was introduced in #40869 . The logic was copied from sections of code that were calculating the bone bounds, but this is not the same calculation needed for `Polygon2D` bounds.
* The canonical software transform for 2D skinning should now be referred to from `rasterizer_canvas_batcher`. This is the most tested path, and a number of bugs were solved there. https://github.com/godotengine/godot/blob/b0c399ec8c9f5f71c64656e8463907153f8459ff/drivers/gles_common/rasterizer_canvas_batcher.h#L1801-L1924
The approach in this PR is _adapted_ from the software skinning (with significant differences).
* The code involved is moved from the h to cpp, as it is quite involved and should be capable of being inlined anyway (just as easily as from the header) as it is called from the same translation unit.
* The old code applied an unused bone `AABB` (0, 0, -1, -1) to the final result even if it was not used. This probably caused little problems due to other oversights, but is strictly speaking incorrect I think, so I did a check for this.
* It is possible that there is a simpler way of doing this. Indeed it can be simpler when the skeleton and `Polygon2D` are guaranteed to be aligned in some ways, but through bug testing the skinning, the robust approach needed the relative transforms.

## Extra
* The same bug is likely in 4.x, as reported in #74656, this PR could be ported.
* The same bug may be present in 3D, we should double check bounds calculations are correct there also if it uses the same approach.

## Discussion
Additionally on rocketchat we discussed 2 additional / alternative methods for skeleton bounds:

1) Single uber-bound that covers the entire skeleton (probably not recalculated)
2) Calculate bounds of the bones _only_, then add a user margin 'fudge factor' to account for the size of the `Polygon2D` _around_ the bones. (This could possibly be derived from the geometry, not sure yet.)

This second approach still has a problem:
We can calculate bone origins in global space (the skeleton in `VisualServer` _does_ have global transform in the form of `skel_base_xform`).
However we need to calculate the `Polygon2D` bound in _local space_. As such we still need a relative transform to go from global to `Polygon2D` local space, probably the global transform for the `Polygon2D` to be updated, which could be expensive. Alternatively we could cull `Polygon2D` in global space directly, but this may not play well with the rest of the engine that expects `get_rect()` to work in local space.

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->

lawnjelly commented:
As suspected with further testing there are more offsets to account for (as in the batching skinning) when the skeleton is offset, but I now have a decent testbed so will see if I can fix these too.

kleonc commented on servers/visual/rasterizer.cpp:
```diff
@@ -561,3 +561,98 @@ int RasterizerStorage::multimesh_get_visible_instances(RID p_multimesh) const {
+		if (l && p_polygon.bones.size() == l * 4 && p_polygon.weights.size() == p_polygon.bones.size()) {
+			// We need to take into account the skeleton base transform
+			// in case it has been offset by the user.
+			Transform2D skel_trans_inv = RasterizerStorage::base_singleton->skeleton_get_base_transform_2d(skeleton).affine_inverse();
+			Transform2D item_transform = skel_trans_inv * xform; // Transform into bone space.
+			Transform2D item_transform_inv = item_transform.affine_inverse(); // Transform back to local space.
```
Note this is equivalent (no need to `affine_inverse()` twice):
```suggestion
			Transform2D skel_trans = RasterizerStorage::base_singleton->skeleton_get_base_transform_2d(skeleton);
			Transform2D item_transform_inv = xform.affine_inverse() * skel_trans;
```

lawnjelly committed:
Fix Polygon2D skinned bounds (for culling)

The bound Rect2 was previously incorrect because bone transforms need to be applied to verts in bone space, rather than local space. This was previously resulting in skinned Polygon2Ds being incorrectly culled.

clayjohn reviewed (approved):
Looks good to me. Tested locally and confirmed it works. This approach is pretty much what we do in 4.x since in 4.x we use Meshes internally when a Polygon2D is used with a Skeleton2D.

I have also started porting this to 4.0.

akien-mga commented:
Thanks!
